### PR TITLE
(fix): subsetting with empty list

### LINKED ISSUE #1149: Subsetting w/ empty list when X is dense throws error
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of anndata.
- [X] (optional) I have confirmed this bug exists on the master branch of anndata.

### Report

Code:

```python
import numpy as np, anndata as ad

ad.AnnData(np.ones((3, 3)))[:, []].X
```

Traceback:

```pytb
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 ad.AnnData(np.ones((3, 3)))[:, []].X

File ~/github/anndata/anndata/_core/anndata.py:656, in AnnData.X(self)
    653     X = None
    654 elif self.is_view:
    655     X = as_view(
--> 656         _subset(self._adata_ref.X, (self._oidx, self._vidx)),
    657         ElementRef(self, "X"),
    658     )
    659 else:
    660     X = self._X

File /usr/local/Cellar/python@3.9/3.9.18/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py:888, in singledispatch.<locals>.wrapper(*args, **kw)
    884 if not args:
    885     raise TypeError(f'{funcname} requires at least '
    886                     '1 positional argument')
--> 888 return dispatch(args[0].__class__)(*args, **kw)

File ~/github/anndata/anndata/_core/index.py:144, in _subset(a, subset_idx)
    142 if all(isinstance(x, cabc.Iterable) for x in subset_idx):
    143     subset_idx = np.ix_(*subset_idx)
--> 144 return a[subset_idx]

IndexError: arrays used as indices must be of integer (or boolean) type
```

Works when X is sparse. Probably not a super common use case.

### Versions

<details>

```
-----
IPython             8.14.0
anndata             0.10.0rc2.dev2+gfa8149d9.d20230927
numpy               1.24.4
session_info        1.0.0
-----
asciitree           NA
asttokens           NA
awkward             2.3.1
awkward_cpp         NA
backcall            0.2.0
cloudpickle         2.2.1
colorama            0.4.6
cython_runtime      NA
cytoolz             0.10.1
dask                2023.9.1
dateutil            2.8.2
decorator           5.1.0
entrypoints         0.3
exceptiongroup      1.0.4
executing           0.8.2
fasteners           0.17.3
google              NA
h5py                3.9.0
importlib_metadata  NA
jedi                0.18.1
jinja2              3.1.1
llvmlite            0.40.1
markupsafe          2.1.1
msgpack             1.0.3
natsort             8.3.0
numba               0.57.1
numcodecs           0.10.2
numexpr             2.8.1
packaging           23.1
pandas              2.1.0
parso               0.8.2
pexpect             4.8.0
pickleshare         0.7.5
prompt_toolkit      3.0.38
psutil              5.9.0
ptyprocess          0.7.0
pure_eval           0.2.1
pyarrow             12.0.1
pygments            2.16.1
pytz                2022.7.1
regex               2.5.112
ruamel              NA
scipy               1.11.2
setuptools_scm      NA
sitecustomize       NA
six                 1.16.0
sparse              0.14.0
sphinxcontrib       NA
stack_data          0.1.4
tblib               1.7.0
tlz                 0.10.1
toolz               0.12.0
traitlets           5.6.0
typing_extensions   NA
wcwidth             0.2.5
xxhash              NA
yaml                5.4.1
zappy               NA
zarr                2.16.1
zipp                NA
zoneinfo            NA
-----
Python 3.9.18 (main, Aug 24 2023, 21:32:48) [Clang 14.0.3 (clang-1403.0.22.14.1)]
macOS-13.4.1-x86_64-i386-64bit
-----
Session information updated at 2023-09-27 12:34
```

</details>

### PR DESCRIPTION
**Pull request by ilan-gold**
- [x] Closes #1149
- [x] Tests added
- [x] Release note added (or unnecessary)

### CODE REVIEW

**Comment by ilan-gold** [comment]
@flying-sheep So you're saying we should catch the error and then report back that the data types are mixed?  Just a side note, this behavior of mixed types doesn't seem to work so this would not be a regression:

```python
from anndata.tests.helpers import gen_adata
adata = gen_adata((100, 100))
adata[[0, 2], :] # good
adata[['cell5'], :] # good
adata[[0, 2, 'cell5'], :] # bad
```

**Comment by ilan-gold** [comment]
Sounds good @flying-sheep

File: anndata/_core/index.py [hunk 1]
```diff
@@ -80,7 +80,16 @@ def name_idx(i):
...
+            dtype = "int"
+            if (
+                all(isinstance(x, str) for x in indexer) and len(indexer) > 0
+            ):  # if not all, but any, then dtype=int will cause an error
+                dtype = "object"
+            elif (
+                any(isinstance(x, str) for x in indexer) and len(indexer) > 0
+            ):  # if not all, but any, there must be more than one type.
+                raise ValueError("Mixed type list indexers not supported.")
+            indexer = np.array(indexer, dtype=dtype)
```
Comment (flying-sheep) [inline anndata/_core/index.py:92]:
```suggestion
        try:
            indexer = np.array(indexer, dtype=dtype)
        except ValueError as e:
            msg = "Mixed type list indexers not supported."
            raise ValueError(msg) from e
```

**Review by flying-sheep — CHANGES_REQUESTED** [review]

**Review by flying-sheep — APPROVED** [review]

**Comment by flying-sheep** [comment]
why no release note?

**Comment by ilan-gold** [comment]
I guess I didn't know what qualified as "unnecessary" for a release note and this seemed like one of those things.

**Comment by flying-sheep** [comment]
Hmm, guess if that's not in the contribution guide we should add it, and if it is, we should link it.

As far as I understand, release notes are not encouraged if an unreleased feature gets modified or if the change is purely development process focused (tests, doc building, ...). This is a bugfix to a live feature, so it should get a release note.

**Review by flying-sheep — CHANGES_REQUESTED** [review]

**Pull request merged**